### PR TITLE
City level data dump handler

### PR DESCRIPTION
**Pull request by cutepig**
- Query accumulated response data by postal code with Athena
- Accumulate postal code level data into city level data
- Push data into `city_level_general_results.json`

### CODE REVIEW

File: src/backend/main.ts [hunk 1]
```diff
@@ -85,41 +86,175 @@ function hash(input: string, pepper: string) {
...
+  ]);
+
+  //
+  // Map query results into the JSON format we want to see in s3
 
   // TODO: Add model for this
-  const data = queryResult.Items[0];
+  const totalResponses = totalResponsesResult.Items[0];
+
+  const cityLevelData = await mapPostalCodeLevelTopostalCodeLevelData(postalCodeLevelDataResponse.Items, bucket);
```
Comment (jareware) [inline src/backend/main.ts:114]:
```suggestion
  const cityLevelData = await mapPostalCodeLevelToPostalCodeLevelData(postalCodeLevelDataResponse.Items, bucket);
```

File: src/backend/main.ts [hunk 2]
```diff
@@ -85,41 +86,175 @@ function hash(input: string, pepper: string) {
...
+
+  //
+  // Push data to S3
+  await s3PutJsonHelper({
+    Bucket: bucket,
+    Key: 'total_responses.json',
+    // TODO: If the cache is same for all JSON, move this to the helper
```
Comment (jareware) [inline src/backend/main.ts:121]: I think this is a safe assumption. :+1:

File: src/backend/main.ts [hunk 3]
```diff
@@ -85,41 +86,175 @@ function hash(input: string, pepper: string) {
...
+          'Population and response count per each city in Finland, where the response count was higher than 25. All the form inputs are coded in city level. Population data from Tilastokeskus. This data is released for journalistic and scientific purposes.',
+        generated: new Date().toISOString(),
+        link: `https://${domain}/city_level_general_results.json`,
+      },
+      data: cityLevelData,
+    },
+  });
+}
+
+async function mapPostalCodeLevelTopostalCodeLevelData(postalCodeLevelData: any[], bucket: string) {
```
Comment (jareware) [inline src/backend/main.ts:150]:
```suggestion
async function mapPostalCodeLevelToPostalCodeLevelData(postalCodeLevelData: any[], bucket: string) {
```

File: src/backend/main.ts [hunk 4]
```diff
@@ -85,41 +86,175 @@ function hash(input: string, pepper: string) {
...
+  const postalCodeCityMappings = await s3GetJsonHelper({ Bucket: bucket, Key: 'postalcode_city_mappings.json' });
+
+  const populationPerCity = await s3GetJsonHelper({ Bucket: bucket, Key: 'population_per_city.json' });
```
Comment (jareware) [inline src/backend/main.ts:153]: Suppose these could also be parallelized..? Though the gains are probably negligible.

File: src/backend/queries.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  COUNT(IF ( stomach_issues = 'no', 1, NULL)) AS stomach_issues_no,
+  COUNT(IF ( stomach_issues = 'yes', 1, NULL)) AS stomach_issues_yes,
+  COUNT(IF ( sensory_issues = 'no', 1, NULL)) AS sensory_issues_no,
+  COUNT(IF ( sensory_issues = 'yes', 1, NULL)) AS sensory_issues_yes,
+  COUNT(IF ( longterm_medication	 = 'no', 1, NULL)) AS longterm_medication_no,
+  COUNT(IF ( longterm_medication	 = 'yes', 1, NULL)) AS longterm_medication_yes,
+  COUNT(IF ( smoking	 = 'no', 1, NULL)) AS smoking_no,
+  COUNT(IF ( smoking	 = 'yes', 1, NULL)) AS smoking_yes,
+  COUNT(IF ( corona_suspicion	 = 'no', 1, NULL)) AS corona_suspicion_no,
+  COUNT(IF ( corona_suspicion	 = 'yes', 1, NULL)) AS corona_suspicion_yes
```
Comment (jareware) [inline src/backend/queries.ts:33]: Eventually these could directly reference fields from our models, to ensure they remain in sync even if changes are made to the model.

Realistically, though, we probably don't want to change them very often, if ever. :smile:

File: src/backend/main.ts (hunk 4 above)
Comment (cutepig) [reply src/backend/main.ts:153]: I think we might be talking milliseconds, though it makes sense. I would even consider parallelizing the whole pipeline for the 2 datasets though, even though I'm not sure about the failure-semantics of `Promise.all`. Meaning, if the other operation rejects in the middle, does it orphan the other that would happily continue? 🤷‍♂️

Comment (jareware) [reply src/backend/main.ts:153]: All tracks within `all()` will finish, but their results are then ignored outside the `all()`. So depends on how you set it up!

But yes, not very important rn. :smile:

**Pull request merged**